### PR TITLE
Add script for handling translations

### PR DESCRIPTION
**Pull request by rffontenelle**
This is an initial take on the enable localization initiative adding a script to ease the babel call.

Usage:

```shell
python babel_runner.py extract           # generate translations template file (.pot)
python babel_runner.py init -l pt_BR     # generate a translation file (.po) for Brazilian Portuguese
python babel_runner.py update            # update all .po files using the existing .pot; or
python babel_runner.py update -l pt_BR   # update only the pt_BR.po
python babel_runner.py compile           # compile all .po files using the existing .pot; or
python babel_runner.py compile -l pt_BR  # compile only the pt_BR.po
```

See #194

### CODE REVIEW

File: babel_runner.py
Comment (hugovk) [inline babel_runner.py:19]: Perhaps prefix the config file with a dot?
```suggestion
MAPPING_FILE = os.path.relpath(os.path.join(PROJECT_DIR, ".babel.cfg"), PROJECT_DIR)
```

**Review by hugovk — COMMENTED** [review]
Thanks for this! Please could you add it somehow to the CI, so we know it at least runs without errors?

I was thinking if this script could be converted into `Makefile` commands, but then we won't have Windows support, so that's no good. But we could add Makefile commands to call this script. This can be left for later though.

File: babel_runner.py
Comment (rffontenelle) [reply babel_runner.py]: Working on a GitHub Actions workflow setup to test the translation files. I renamed the mapping file to `.babel.cfg`. By the way, I was about to go crazy with babel\<tab> completing with both mapping file and script instead of only the script, so thanks for the suggestion!

**Comment by rffontenelle** [comment]
@hugovk workflow job added to tests.yml

File: .github/workflows/tests.yml
Comment (Wulian233) [inline .github/workflows/tests.yml:74]: Here you are using python3 and using 

> os: ["ubuntu-latest", "windows-latest"] 

to run on multiple systems. But python3 doesn't work on my Windows, use python?

Comment (rffontenelle) [reply .github/workflows/tests.yml:74]: That won't be necessary, as [actions/setup-python](https://github.com/actions/setup-python) installs python3 binary (or symlink to binary, I don't know), so those commands work fine in the CI environment. See [this Windows workflow run in my fork](https://github.com/rffontenelle/python-docs-theme/actions/runs/10221587570/job/28284437047).

Any documentation we might write will probably need to take your scenario in consideration.

Comment (Wulian233) [reply .github/workflows/tests.yml:74]: Great🚀

Comment (AA-Turner) [inline .github/workflows/tests.yml:74]: ``python`` is more typical in GHA (& consistent with other jobs in this file). I don't think setuptools is needed?
```suggestion
          pip install -U pip
          pip install Babel jinja2
      - run: python babel_runner.py extract
      - run: python babel_runner.py init -l pt_BR
      - run: python babel_runner.py update
      - run: python babel_runner.py update -l pt_BR
      - run: python babel_runner.py compile
      - run: python babel_runner.py compile -l pt_BR
```

File: babel_runner.py
Comment (AA-Turner) [inline babel_runner.py:9]:
```suggestion
import argparse
import os
import subprocess
import tomllib
```

File: babel_runner.py
Comment (AA-Turner) [inline babel_runner.py:21]: It might be nicer to use `pathlib` here? Also, any reason to use relative paths rather than absolute?

File: .github/workflows/tests.yml
Comment (rffontenelle) [reply .github/workflows/tests.yml:74]: Installing babel and jinja2 without setuptools made extraction fail with `Unknown extraction method 'jinja2' babel`. I even reproduced locally and only solved after a pip's --force-reinstall

Comment (AA-Turner) [reply .github/workflows/tests.yml:74]: Oh, I remember -- `babel` uses setuptools' `Command` class internally. Hence the mess that is ``babel_runner.py`` in Sphinx...

A

File: babel_runner.py
Comment (rffontenelle) [reply babel_runner.py:9]: I added this blank line because of ruff linter: https://github.com/python/python-docs-theme/actions/runs/10208999238/job/28250436250

Comment (AA-Turner) [reply babel_runner.py:9]: Try setting python-version for Ruff -- it is likely defaulting to something before 3.11

Comment (rffontenelle) [reply babel_runner.py:9]: You are correct. Ruff uses python-version from pyproject.toml, which is currently set to ">= 3.8".

Should I set ruff's target-version permanently, patch it for the workflow environment, or is there another way of passing the Python version to ruff via CLI?

Comment (AA-Turner) [reply babel_runner.py:9]: https://docs.astral.sh/ruff/configuration/#the-config-cli-flag

``--config "target-version=py311"``

File: .github/workflows/tests.yml
Comment (hugovk) [reply .github/workflows/tests.yml:74]: We already use `python` rather than `python3` in _this_ repo, so I suggest we keep it the same here.

Comment (rffontenelle) [reply .github/workflows/tests.yml]: Ok, change applied. It looks like there are 3 votes on `python` (AA-Turner too) :D

File: babel_runner.py
Comment (rffontenelle) [reply babel_runner.py]: I was using relative path because the output would be too long because of the absolute path. By using relative paths I achieved `creating catalog locales/pt_BR/LC_MESSAGES/messages.po based on locales/messages.pot` (just 'locales/...' instead of the absolute path).

But using the feedback from both of you, I replaced os with pathlib and used `cwd` keyword argument of `subprocess.run()` in a way that this whole relative path pollution is no longer necessary.

File: babel_runner.py
Comment (rffontenelle) [reply babel_runner.py]: Well, the problem is that ruff check is failing is not just on my end, but in lint.yml action run, so AFAIK I'm not able to set the python version unless is replace python-version or set target-version in pyproject.toml.

I've worked around this issue using tomli as alternative (used your suggestion in babel add TOML support pull request). For this I added requirements.txt to use as single point of installing tomli or not based on Python version.

**Review by hugovk — COMMENTED** [review]
A few nits, then I think we can merge this and iterate if needed, unless anyone else has any last comments?

**Comment by Wulian233** [comment]
In the Simplified Chinese documentation, there is an issue regarding hard-coded strings such as "added in version" and "deprecated in version" etc. that cannot be translated. Can this PR solve this problem?

https://github.com/python/python-docs-zh-cn/issues/364

**Comment by rffontenelle** [comment]
> In the Simplified Chinese documentation, there is an issue regarding hard-coded strings such as "added in version" and "deprecated in version" etc. that cannot be translated. Can this PR solve this problem?
> 
> https://github.com/python/python-docs-zh-cn/issues/364

No, that's a Sphinx string and it is not translated to zh_CN: https://github.com/sphinx-doc/sphinx/blob/116a430caca98580c8877cf7b0ba270eae9604bf/sphinx/locale/zh_CN/LC_MESSAGES/sphinx.po#L1951

Consider contributing in https://explore.transifex.com/sphinx-doc/sphinx-1/

File: babel_runner.py
Comment (ezio-melotti) [inline babel_runner.py:77]:
```suggestion
    if locale:
```

File: babel_runner.py
Comment (ezio-melotti) [inline babel_runner.py:85]:
```suggestion
    if locale:
```

File: babel_runner.py
Comment (ezio-melotti) [inline babel_runner.py:109]:
```suggestion
        if not locale:
```

File: babel_runner.py
Comment (ezio-melotti) [inline babel_runner.py:104]:
```suggestion
    parser.add_argument(
        "-l",
        "--locale",
        default="",
        help="language code (needed for init, optional for update and compile)",
    )

    args = parser.parse_args()
    locale = args.locale
```
This should be equivalent.

**Review by hugovk — APPROVED** [review]
Thank you!

**Comment by Wulian233** [comment]
It looks great! I have a question: where should we translate this theme, and how will the translation automatically sync to this repository?

**Comment by rffontenelle** [comment]
@Wulian233 I mentioned that in the issue and it is yet to be defined.

Some projects coordinate translation efforts in the own forge via pull/merge requests — not optimal for the translators. There is also translation platforms like Weblate, Transifex and others.

My vote would be to use Weblate, which provides "libre hosting" for free software and features automatic opening and updating pull request for contributed translations. Tasks involve: setting up Weblate project (I can help with/do it), maintainers here update POT and give time before releasing to receive translations; and merging translation pull requests from Weblate.

**Pull request merged**
